### PR TITLE
NF-844 - Cache Money v1

### PR DESCRIPTION
**Pull request by alexj-novisto**

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+sys-deps:
+	pip install -U pre-commit poetry==1.0.10 tox
+	pre-commit install
+
+install:
+	poetry install
+
+run:
+	poetry run python run.py
+
```
Comment (MarcDufresne) [inline Makefile:14]: There's no `run.py`

File: Makefile [hunk 2]
```diff
@@ -0,0 +1,31 @@
+default:
+	@echo "See usage in Makefile"
+
+.PHONY: sys-deps install run format lint tests build
+sys-deps:
+	pip install -U pre-commit poetry==1.0.10 tox
```
Comment (MarcDufresne) [inline Makefile:6]: Missing tox-docker

File: Makefile [hunk 3]
```diff
@@ -0,0 +1,31 @@
+default:
+	@echo "See usage in Makefile"
+
+.PHONY: sys-deps install run format lint tests build
```
Comment (MarcDufresne) [inline Makefile:4]: You'll have to remove `run` form here too

File: LICENSE [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+      To apply the Apache License to your work, attach the following
+      boilerplate notice, with the fields enclosed by brackets "[]"
+      replaced with your own identifying information. (Don't include
+      the brackets!)  The text should be enclosed in the appropriate
+      comment syntax for the file format. We also recommend that a
+      file or class name and description of purpose be included on the
+      same "printed page" as the copyright notice for easier
+      identification within third-party archives.
+
+   Copyright [yyyy] [name of copyright owner]
```
Comment (MarcDufresne) [inline LICENSE:189]: I guess you need to fill this?

File: .gitignore [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+[Dd]esktop.ini
+$RECYCLE.BIN/
+*.cab
+*.msi
+*.msix
+*.msm
+*.msp
+*.lnk
+local.yml
+unit_test.yml
```
Comment (MarcDufresne) [inline .gitignore:125]: I guess we don't have a config on this repo, so probably don't need this

Comment (MarcDufresne) [reply .gitignore:125]: Or we do for the tests? 🤔

File: README.md [hunk 6]
```diff
@@ -0,0 +1,147 @@
+# Cache Money
+
+[![Coverage Status](https://coveralls.io/repos/github/novisto/cache-money/badge.svg)](https://coveralls.io/github/novisto/cache-money)
+
+Async cache library for [memoization](https://en.wikipedia.org/wiki/Memoization) using Redis. Inspired by 
+[Walrus](https://github.com/coleifer/walrus) and implemented with aioredis.
```
Comment (MarcDufresne) [inline README.md:6]: trivial: you linked to Walrus, maybe link to aioredis as well

File: README.md [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+execution of the function will be skipped.
+
+You can add a timeout in the declaration of the decorator, you can find constants for common timeout duration in 
+`cache_money.constants.py`. When the timeout is reached, Redis remove the entry itself.
+
+It's also possible to clear the cache early by using the method bust that gets added to a function decorated by 
+Cache Money. An example is provided below.
+
+This library is available on PyPI under the name cache-money. You can install with pip by running pip install 
+cache-money.
```
Comment (MarcDufresne) [inline README.md:20]:
```suggestion
This library is available on PyPI under the name cache-money. You can install with pip by running 
`pip install cache-money`.
```

File: README.md [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+
+# Requirements
+
+You need a redis instance running to use this library. This library was tested to run on version of Redis >= 4.0.0. 
+If you have docker set up you can create a redis instance like this:
+
+```
+docker run -d -p 6379:6379 redis:6.2.5
+```
+
```
Comment (MarcDufresne) [inline README.md:31]: Maybe give an example with the `make redis-start` as well

File: Makefile [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+	poetry run isort $(LINT_TARGETS)
+
+lint:
+	poetry run flake8 $(LINT_TARGETS)
+
+tests:
+	poetry run pytest --cov-report term-missing --cov cache_money/ tests/
+
+redis-start:
+	docker run -d -p 63798:6379 --name cache_money redis:6.2.5
```
Comment (MarcDufresne) [inline Makefile:28]: trivial: maybe `cache_money_redis` for the name (and in the `stop` as well)

File: README.md [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+
+## Contributing and getting set up for local development
+
+To set yourself up for development on Cache Money, make sure you are using
+[poetry](https://poetry.eustace.io/docs/) and simply run the following commands from the root directory:
+
+```bash
+make bootstrap
```
Comment (MarcDufresne) [inline README.md:145]:
```suggestion
make sys-deps
```

File: azure-pipelines.yml [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+      - master
+
+trigger:
+  branches:
+    include:
+      - master
+
+variables:
+  dockerRegistryServiceConnection: '0718fd53-7517-42de-9ace-096cfbd39374'
+  containerRegistry: 'novisto.azurecr.io'
```
Comment (MarcDufresne) [inline azure-pipelines.yml:15]: No need since we don't publish a Docker image

File: azure-pipelines.yml [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+      vmImage: $(vmImageName)
+    jobs:
+      - job: Tests
+        displayName: Lint and Tests
+        steps:
+          - task: Docker@2
+            displayName: Login to ACR
+            inputs:
+              command: login
+              containerRegistry: $(dockerRegistryServiceConnection)
```
Comment (MarcDufresne) [inline azure-pipelines.yml:31]: No need of that either

File: azure-pipelines.yml [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+          - script: make sys-deps
+            displayName: Install sys deps
+          - task: Cache@2
+            inputs:
+              path: .tox
+              key: 'v1 | tox | poetry.lock'
+              restoreKeys: |
+                v1 | tox
+          - script: poetry config http-basic.novisto $(PYPI_NOVISTO_USERNAME) $(PYPI_NOVISTO_PASSWORD)
+            displayName: Configure Poetry auth
```
Comment (MarcDufresne) [inline azure-pipelines.yml:45]: We're not using private packages so we don't need that

File: azure-pipelines.yml [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+          - script: tox -e install
+            displayName: Install project deps
+          - script: tox -e lint
+            displayName: Lint
+          - script: tox -e tests
+            displayName: Run tests
+          # TODO: Add Coverage
```
Comment (MarcDufresne) [inline azure-pipelines.yml:52]: We'll need to add coveralls publish here

File: .gitignore (hunk 5 above)
Comment (alexj-novisto) [reply .gitignore]: no opset/config

File: LICENSE (hunk 4 above)
Comment (alexj-novisto) [reply LICENSE]: switched to MIT as discussed, filled the same in MIT

**Review by MarcDufresne — APPROVED** [review]

File: cache_money/cache.py [hunk 15]
```diff
@@ -0,0 +1,288 @@
...
+            timeout: Time to cache return values in seconds.
+
+        Returns:
+            Return the result of the decorated function call with the given args and kwargs.
+        """
+
+        def decorator(fn):
+            def make_key(args, kwargs):
+                """Apply the name of the function in front of the key, separated by a colon `:`."""
+                return f"{fn.__name__}:{key_fn(args, kwargs)}"
```
Comment (mdupuisnovisto) [inline cache_money/cache.py:209]: A function's name is not really a signature. Many functions can have the same name. I would suggest putting the function's module (maybe part of it depending on how long the key can be) part of they key.

If we decide not to make the change, I suggest we document that.

**Review by mdupuisnovisto — APPROVED** [review]

File: pyproject.toml [hunk 16]
```diff
@@ -2,7 +2,7 @@
 name = "cache-money"
 version = "1.0.0"
 description = "Async cache decorator for memoization using aioredis."
-authors = ["Alexandre Jutras <alexandre.jutras@novisto.com>"]
+authors = ["Alexandre Jutras <alexandre.jutras@novisto.com>", "Marc-André Dufresne <marc-andre.dufresne@novisto.com"]
```
Comment (MarcDufresne) [inline pyproject.toml:5]: 🙏

**Pull request merged**